File: NeedleHaystack.java
// Time Complexity : O(m + n)
// Space Complexity : O(n)

class Solution {
    public int strStr(String haystack, String needle) {
        if(needle == null || needle.length() == 0)
            return 0;
        if(haystack == null || haystack.length() == 0)
            return -1;
        int i = 0;
        int j = 0;
        int[] lps = helper(needle); 
        while(i < haystack.length()){
            if(haystack.charAt(i) == needle.charAt(j)){
                j++;
                i++;
                if(j == needle.length()){
                    return i - needle.length();
                }
            }
            else if(haystack.charAt(i) != needle.charAt(j) && j > 0){
                j = lps[j - 1];
            }
            else if(haystack.charAt(i) != needle.charAt(j) && j == 0){
                i++;
            }
        }
        return -1;
    }
    
    private int[] helper(String needle){
        int[] lps = new int[needle.length()];
        int i = 1;
        int j = 0;
        while(i < needle.length()){
            if(needle.charAt(i) == needle.charAt(j)){
                j++;
                lps[i] = j;
                i++;
            }
            else if(needle.charAt(i) != needle.charAt(j) && j > 0){
                j = lps[j - 1];
            }
            else if(needle.charAt(i) != needle.charAt(j) && j == 0){
                lps[i] = 0;
                i++;
            }
        }
        return lps;
    }
}

File: needle_haystack.py
# Time Complexity :
# O(NM)  - Size of the needle * size of the haystack

# Space Complexity :
# O(1) - There is no auxillary space being allocated

# Did this code successfully run on Leetcode :
#Yes

#We store a needle index to 0
#We go through each character of the haystack and check if it's character is same as the character in needle. If it is, we go ahead to next character in both, if it's not, we set the needle index to 0 and haystack index to where the needle started matching
#We do this until haystack index reaches the end and if anywhere on the way, needle index reaches its end, we return the needle start index

class Solution:
    def strStr(self, haystack: str, needle: str) -> int:

        if len(haystack) == 0 or len(needle) == 0:
            return -1
        j = 0
        needle_found_start = -1
        i= 0
        #for i,char in enumerate(haystack):
        while i in range(len(haystack)):
            char = haystack[i]
            if char == needle[j] :
                if j == 0 :
                    needle_found_start = i
                j += 1
            else :
                if needle_found_start != -1 :
                    i = needle_found_start
                needle_found_start = -1
                j = 0

            if j == len(needle):
                return needle_found_start

            i += 1

        return -1
